language/move-binary-format/src/file_format.rs

```diff
@@ -666,10 +666,28 @@ impl AbilitySet {
     /// For a polymorphic type, its actual abilities correspond to its declared abilities but
     /// predicated on its non-phantom type arguments having that ability. For `Key`, instead of needing
     /// the same ability, the type arguments need `Store`.
-    pub fn polymorphic_abilities(
+    pub fn polymorphic_abilities<S, T>(
         declared_abilities: Self,
-        type_arguments: impl IntoIterator<Item = (Self, bool)>,
-    ) -> Self {
+        declared_phantom_parameters: S,
+        type_arguments: T,
+    ) -> PartialVMResult<Self>
+    where
+        S: IntoIterator<Item = bool>,
+        T: IntoIterator<Item = Self>,
+        S::IntoIter: ExactSizeIterator,
+        T::IntoIter: ExactSizeIterator,
+    {
+        let declared_phantom_parameters = declared_phantom_parameters.into_iter();
+        let type_arguments = type_arguments.into_iter();
+
+        if declared_phantom_parameters.len() != type_arguments.len() {
+            return Err(
+                PartialVMError::new(StatusCode::VERIFIER_INVARIANT_VIOLATION).with_message(
+                    "the number of arguments doesn't match the number of parameters".to_string(),
```

Comment:
maybe something a tad more specific with regards to `declared_phantom_parameters` not matching the length of `type_arguments`